go/vt/vtgate/plan_execute.go

```diff
@@ -79,21 +79,28 @@ func (e *Executor) newExecute(
 
 	query, comments := sqlparser.SplitMarginComments(sql)
 
-	// 2: Parse and Validate query
+	// 2: Parse and Validate query.
 	stmt, reservedVars, err := parseAndValidateQuery(query, e.env.Parser())
 	if err != nil {
 		return err
 	}
 
-	var lastVSchemaCreated time.Time
-	vs := e.VSchema()
-	lastVSchemaCreated = vs.GetCreated()
+	var (
+		vs                 = e.VSchema()
+		lastVSchemaCreated = vs.GetCreated()
+		result             *sqltypes.Result
+		plan               *engine.Plan
+	)
+
 	for try := 0; try < MaxBufferingRetries; try++ {
-		if try > 0 && !vs.GetCreated().After(lastVSchemaCreated) {
+		if try > 0 && !vs.GetCreated().After(lastVSchemaCreated) { // We need to wait for a vschema update
 			// There is a race due to which the executor's vschema may not have been updated yet.
-			// Without a wait we fail non-deterministically since the previous vschema will not have the updated routing rules
-			if waitForNewerVSchema(ctx, e, lastVSchemaCreated) {
+			// Without a wait we fail non-deterministically since the previous vschema will not have
+			// the updated routing rules.
+			timeout := e.resolver.scatterConn.gateway.buffer.GetConfig().MaxFailoverDuration / MaxBufferingRetries
```

Comment:
could you explain why this timeout is calculated in this way?

Reply:
We retry 3 times before giving up. How long we wait before each retry was previously hardcoded at 30s while the min buffer window defaults to 20s. I thought it made more sense to instead base the wait time on how long the wait time is.

My thinking for the calculation was that this way we should be able to perform the max retries within the given window of time for many queries (certainly not all) and we should not end up waiting too long after the buffer window has ended, retrying old queries. 

Before we could end up waiting 90 seconds to retry 3 times and continuing to retry when the buffering window ended 60+ seconds ago. That's of course assuming that for some reason we didn't receive a newer vschema.

Thinking about it again... I think this improves some failure scenarios/conditions but doesn't make the typical happy path much better as there's no point in retrying when we haven't gotten a new vschema UNLESS the traffic switch failed and we backed out and those queries should now succeed with the original targets/routing. Let's discuss tomorrow. I'm happy to change this... and should add a comment either way. 🙂

Reply:
Noting that we discussed this directly and were both OK with the calculation. I will add a comment, however, to describe the reasoning (which I should have already done).